[TIL-95] 기술학습 리스트 구현 페이징 기능 추가 및 테스트

dongyeon-0822 opened the pull request:
## 이슈 번호(링크)

[TIL-95](https://soma-til.atlassian.net/browse/TIL-95)

<br>

## 개요

- 기술학습 리스트 구현 페이징 기능 추가 및 테스트

<br>

## 내용

- Category, ProblemCategory 엔티티 추가
- Problem 엔티티 컬럼 수정 및 추가
- 페이징 기능 추가
- 리스트 정상적 반환 로직 테스트 

<br>

## 리뷰어한테 할 말

- 문제 & 카테고리 DB ERD 수정: [DB 문서](https://soma-til.notion.site/DB-e8c049fe0c2d49e4b673625d280c06f6?pvs=4) 
- PageDto를 만들었다가 사용되는 dto가 2개 밖에 없어서 로직만 복잡해진 것 같아 없앴는데, 다시 생각해보니 있는게 나을 것 같기도 하네요. 
- 객체 변환 로직은 비즈니스 로직이 아니라고 생각해서 서비스가 아닌 Dto of함수에 작성하였는데 이렇게 작성해도 괜찮을지 피드백 주시면 감사하겠습니다!
- JPA N+1 문제 발생할 것을 예상하고 있지만, 처음부터 고려하여 구현하기보다 추후 개선하는게 나을 것 같아 여기서는 고려하지 않았습니다..!


[TIL-95]: https://soma-til.atlassian.net/browse/TIL-95?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

dongyeon-0822 committed:
feat: 기술학습 문제 리스트 컬럼 추가 및 페이징 기능 구현

- 카테고리 관련 로직 구현
- 페이징 기능 추가

dongyeon-0822 committed:
feat: 기술학습 문제 리스트 서비스 로직 테스트

- 문제 리스트를 정상적으로 반환하는지 테스트

dongyeon-0822 committed:
feat: 기술학습 문제 리스트 예외처리 로직 추가

CodeDiary18 commented on til-api/src/main/java/com/til/controller/problem/ProblemController.java:
```diff
@@ -21,8 +23,14 @@ public class ProblemController {
-    public ApiResponse<ProblemListResponse> getProblemList() {
-        List<ProblemListDto> problemList = problemService.getProblemList();
-        return ApiResponse.ok(ProblemListResponse.of(problemList));
+    public ApiResponse<ProblemListResponse> getProblemList(
+        @RequestParam(defaultValue = "0") int page,
+        @RequestParam(defaultValue = "10") int size) {
+
+        if (page < 0 || size <= 0) {
+            throw new BaseException(ProblemErrorCode.INVALID_PAGE_REQUEST);
+        }
```
P3: PageRequest dto를 생성해서 다른 페이징이 필요한 부분에서도 재활용할 수 있도록 적용해보면 좋을 것 같습니다!

dongyeon-0822 replied:
넵! 적용하는게 더 좋을 것 같습니다!

CodeDiary18 commented on til-api/src/main/java/com/til/controller/problem/reponse/ProblemListResponse.java:
```diff
@@ -3,12 +3,26 @@
 public record ProblemListResponse(
-                                  List<ProblemListDto> problemList
+                                  List<ProblemListDto> problemList,
+                                  int page,
+                                  int size,
+                                  long totalElements,
+                                  int totalPages
```
P3: 이 부분도 한번에 PageInfoDto 같은 걸로 묶어서 던지면 좋을 것 같습니다!

dongyeon-0822 replied:
이 부분도 PageDto로 묶어보겠습니다!

CodeDiary18 reviewed (approved):
저의 전반적인 의견은 페이징 관련 로직 재사용을 위한 dto 리팩토링 요청입니다!  
페이징이 재사용 수요가 많은 부분이라 별도 task로 분리하여 리팩토링 작업해도 괜찮을 것 같고 이번 task에 피드백 반영해도 좋을 것 같습니다😀

doburn commented on til-domain/src/main/java/com/til/domain/problem/enums/ProblemErrorCode.java:
```diff
@@ -0,0 +1,19 @@
+
+@Getter
+@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
+public enum ProblemErrorCode implements ErrorCode {
+
+    INVALID_PAGE_REQUEST(HttpStatus.BAD_REQUEST, "잘못된 페이지 요청입니다.");
+
+    private final HttpStatus status;
+    private final String message;
+}
```
P3: 잘못된 페이지 요청 코드는 다른곳에서도 재활용할 여지가 있어보여서 `ProblemErrorCode` 보다 `PageErrorCode` 같은 이름으로 공통 에러코드로 빼는건 어떠신가요?

dongyeon-0822 replied:
넵 좋은 의견같습니다! 수정해보겠습니다!

doburn commented on til-domain/src/main/java/com/til/domain/category/model/ProblemCategory.java:
```diff
@@ -0,0 +1,34 @@
+@Getter
+@SuperBuilder
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Table(name = "problem_category")
+public class ProblemCategory extends BaseTimeEntity {
+
+    @Id
+    private Long id;
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "problem_id")
+    private Problem problem;
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "category_id")
+    private Category category;
+}
```
P5: 사실 이 부분은 제 초기 ERD 설계를 바탕으로 문제와 카테고리의 M:N 관계를 구현하신 코드지만, 1:N으로 만드는게 맞나 계속 고민이 되네요

dongyeon-0822 replied:
저도 이 부분에 대한 고민이 많았는데, 나중에 멘토님 의견듣고 리팩토링하면 좋을 것 같습니다!

doburn reviewed (commented):
초기 코드라 고민이 많으셨을 것 같습니다!

dongyeon-0822 committed:
refactor: 기술학습 문제 리스트 리팩토링

- domain 모듈: PageInfoDto 페이지 정보 묶어 반환 처리
- api 모듈: PageParamRequest, PageResponse 생성, ProblemListResponse 삭제
- ProblemErrorCode -> PageErrorCode 로 공통 에러 코드에 추가
- 파일명 수정: ProblemDto -> ProblemInfoDto, ProblemListDto -> ProblemListInfoDto, ProblemPageDto -> ProblemListDto
